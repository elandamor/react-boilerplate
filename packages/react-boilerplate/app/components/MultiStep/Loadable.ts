/**
 *
 * Asynchronously loads the component for MultiStep
 *
 */

import Loadable from 'react-loadable';

export default Loadable({
  loader: () => import('./index'),
  loading: () => null,
});
