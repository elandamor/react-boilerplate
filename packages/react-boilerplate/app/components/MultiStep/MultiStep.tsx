import React, { Component } from 'react';
import classNames from 'classnames';
// Components
import Button from '../Button';
// Styles
import Wrapper, { Actions, Tracker } from './styles';

// import { makeDebugger } from '../../lib';
// const debug = makeDebugger('MultiStep');

/**
 * @render react
 * @name MultiStep container
 * @description MultiStep container.
 * @example
 * const steps = [
 *  { name: 'StepOne', component: <StepOne /> },
 *  { name: 'StepTwo', component: <StepTwo /> },
 *  { name: 'StepThree', component: <StepThree /> },
 * ];
 *
 * <MultiStep steps={steps} />
 */

interface IProps {
  className?: string;
  onSubmit: () => void;
  steps: Array<{
    component: JSX.Element;
    name: string;
  }>;
}

interface IState {
  [key: string]: any;
}

const getNavStates = (indx, length) => {
  const styles = [];

  for (let i = 0; i < length; i++) {
    if (i < indx) {
      styles.push('done');
    } else if (i === indx) {
      styles.push('doing');
    } else {
      styles.push('todo');
    }
  }
  return { current: indx, styles };
};

const checkNavState = (currentStep, stepsLength) => {
  if (currentStep > 0 && currentStep < stepsLength - 1) {
    return {
      showDoneBtn: false,
      showNextBtn: true,
      showPreviousBtn: true,
    };
  } else if (currentStep === 0) {
    return {
      showDoneBtn: false,
      showNextBtn: true,
      showPreviousBtn: false,
    };
  } else {
    return {
      showDoneBtn: true,
      showNextBtn: false,
      showPreviousBtn: true,
    };
  }
};

// eslint-disable-next-line react/prefer-stateless-function
class MultiStep extends Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);

    this.state = {
      currentStep: 0,
      navState: getNavStates(0, this.props.steps.length),
      showNextBtn: true,
      showPreviousBtn: false,
    };
  }

  public setNavState = (next) => {
    this.setState({
      navState: getNavStates(next, this.props.steps.length),
    });
    if (next < this.props.steps.length) {
      this.setState({ currentStep: next });
    }
    this.setState(checkNavState(next, this.props.steps.length));
  }

  public handleKeyDown = (event) => {
    if (event.which === 13) {
      this.next();
    }
  }

  public handleClick = (event) => {
    const { steps } = this.props;

    if (
      event.currentTarget.value === steps.length - 1 &&
      this.state.currentStep === steps.length - 1
    ) {
      this.setNavState(steps.length);
    } else {
      this.setNavState(event.currentTarget.value);
    }
  }

  public next = () => {
    this.setNavState(this.state.currentStep + 1);
  }

  public previous = () => {
    const { currentStep } = this.state;

    if (currentStep > 0) {
      this.setNavState(currentStep - 1);
    }
  }

  public renderSteps = () => {
    const { navState } = this.state;

    return this.props.steps.map((s, i) => (
      <li
        className={classNames('c-step', ` -${navState.styles[i]}`)}
        onClick={this.handleClick}
        key={i}
        value={i}
      >
        <i className="c-icon" />
        <span>{this.props.steps[i].name}</span>
      </li>
    ));
  }

  public render() {
    const { className } = this.props;

    return (
      <Wrapper
        className={classNames('', className)}
        onKeyDown={this.handleKeyDown}
      >
        <Tracker>{this.renderSteps()}</Tracker>
        {this.props.steps[this.state.currentStep].component}
        <Actions>
          <Button onClick={this.previous}>Previous</Button>
          {!this.state.showDoneBtn ? (
            <Button onClick={this.next}>Next</Button>
          ) : (
            <Button onClick={() => this.props.onSubmit()}>Done</Button>
          )}
        </Actions>
      </Wrapper>
    );
  }
}

export default MultiStep;
