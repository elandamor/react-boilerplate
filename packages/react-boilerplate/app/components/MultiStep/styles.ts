import styled from 'styled-components';

const Wrapper = styled.div`
  padding: 24px 0;
  position: relative;
`;

const Tracker = styled.ol`
  align-items: center;
  display: flex;
  justify-content: space-between;
  margin: 0;
  padding: 0;
  width: 100%;

  li {
    align-items: center;
    cursor: pointer;
    display: flex;
    flex: 1;
    flex-direction: column;
    justify-content: center;
    list-style-type: none;
    position: relative;

    .c-icon {
      border: 4px solid red;
      border-radius: 100%;
      display: block;
      height: 48px;
      margin-bottom: 12px;
      width: 48px;
    }

    &.-doing,
    &.-done {
      .c-icon {
        background-color: red;
      }
    }

    &.-doing,
    &.-done {
      &:not(:first-child):before {
        background-color: red;
        content: '';
        height: 4px;
        left: -50%;
        margin-top: -2px;
        position: absolute;
        top: 24px;
        width: 100%;
      }
    }
  }
`;

const Actions = styled.div``;

export default Wrapper;

export { Actions, Tracker };
