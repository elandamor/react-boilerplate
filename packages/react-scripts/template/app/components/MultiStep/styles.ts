import styled, { css } from 'styled-components';

const Wrapper = styled.div``;

export const Navigation = styled.div`
  display: flex;
  justify-content: center;
`;

export const Tracks = styled.ol`
  display: flex;
  justify-content: center;
  margin: 0;
  padding: 40px 0;

  ${(props: { verticalTrack: boolean }) => props.verticalTrack && css`
    align-items: flex-start;
    flex-direction: column;
  `};
`;

export const Track = styled.li`
  align-items: ${
    (props: { verticalTrack: boolean }) => props.verticalTrack
    ? 'flex-start' : 'center'
  };
  display: flex;
  flex-direction: column;
  justify-content: center;
  min-width: 160px;
  position: relative;

  .side-nav {
    &__wrap {
      align-items: center;
      cursor: pointer;
      display: flex;
      flex-direction: ${
        (props: { verticalTrack: boolean }) => props.verticalTrack
        ? 'row' : 'column'
      };
      justify-content: center;
      min-height: 40px;
    }

    &__circle {
      align-items: center;
      background-color: #eeeeee;
      border-radius: 50%;
      display: flex;
      height: 40px;
      justify-content: center;
      margin: 20px;
      width: 40px;

      &-inner {
        background-color: #ffffff;
        border-radius: 50%;
        display: block;
        height: 30px;
        width: 30px;
      }
    }

    &__line {
      background-color: #c0d0f3;
      border-radius: 2px;
      height: 4px;
      position: absolute;
      right: -25%;
      visibility: hidden;
      width: 50%;

      ${(props: { verticalTrack: boolean }) => props.verticalTrack && css`
        bottom: -12.5%;
        height: 25%;
        width: 4px;
        right: auto;
      `};
    }

    &__page {
      display: none;
    }
  }

  &.-done {
    .side-nav__circle {
      background-color: #c0d0f3;
    }
  }

  &.-doing {
    .side-nav__circle {
      background-color: #ffc73d;
      background-image: linear-gradient(45deg,#ffc73d,#ff722a 99%);
    }

    .side-nav__page {
      font-weight: 800;
    }
  }

  &:not(:last-child).-done {
    .side-nav__line {
      visibility: visible;
    }
  }

  &:hover {
    .side-nav__circle {
      background-color: #d5d5d5;
    }

    .-done {
      .side-nav__circle {
        background-color: #d5d5d5;
      }
    }
  }
`;

export default Wrapper;
