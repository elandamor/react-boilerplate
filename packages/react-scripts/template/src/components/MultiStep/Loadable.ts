/**
 *
 * Asynchronously loads the component for MultiStep
 *
 */

import React from 'react';

export default React.lazy(() => import('./index'));
