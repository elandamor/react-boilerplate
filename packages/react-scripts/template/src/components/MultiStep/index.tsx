import React, { FC, useState } from 'react';
import classNames from 'classnames';
// Styles
import Wrapper, { Navigation, Track, Tracks } from './styles';
import Button from '../Button';
import { KEYBOARD_CODE } from '../../constants';

// import { makeDebugger } from '../../utils';
// const debug = makeDebugger('MultiStep');

interface IProps {
  className?: string;
  doneText?: string;
  onDone?: () => void;
  showNavigation?: boolean;
  showNavigationTop?: boolean;
  steps: {
    name: string,
    component: React.ReactNode,
  }[];
  verticalTrack?: boolean;
};

/**
 * @render react
 * @name MultiStep component
 * @description MultiStep component.
 * @todo Make code more readable
 * @example
 * <MultiStep
 *    steps={[
 *      { name: 'Name', component: <Step /> },
 *      { name: 'Email', component: <Step /> }
 *    ]}
 * />
 */

const MultiStep: FC<IProps> = ({
  className,
  doneText,
  onDone: handleDone,
  showNavigation,
  showNavigationTop,
  steps,
  verticalTrack,
}) => {
  const [styles, setStyles] = useState(getNavStyles(0, steps.length))
  const [compState, setComp] = useState(0)
  const [buttonsState, setButtons] = useState(getButtonsState(0, steps.length))

  const setStepState = (index: number) => {
    setStyles(getNavStyles(index, steps.length))
    setComp(index < steps.length ? index : compState)
    setButtons(getButtonsState(index, steps.length))
  }

  const next = (stepsLength?: number) => setStepState(compState + 1)

  const previous = () => setStepState((compState > 0) ? compState - 1 : compState)

  const handleClick = (event: React.BaseSyntheticEvent) => {
    if (event.currentTarget.value === steps.length - 1
      && compState === steps.length - 1) {
      setStepState(steps.length)
    } else {
      setStepState(event.currentTarget.value)
    }
  }

  const handleKeyDown = (event: React.KeyboardEvent) => {
    return event.which === KEYBOARD_CODE.ENTER ? next(steps.length) : {};
  }

  const renderTracks = () => {
    return (
      // @ts-ignore
      <Tracks verticalTrack={verticalTrack}>
        {
          steps.map((_: any, index: number) => (
            <Track
              className={styles[index]}
              onClick={handleClick}
              key={index}
              value={index}
              // @ts-ignore
              verticalTrack={verticalTrack}
            >
              <div className="side-nav__wrap">
                <div className="side-nav__circle">
                  <div className="side-nav__circle-inner" />
                  <div className="side-nav__line" />
                </div>
                <span className="side-nav__page">
                  {steps[index].name}
                </span>
              </div>
            </Track>
          ))
        }
      </Tracks>
    )
  }

  const renderNavigation = () => {
    return showNavigation && (
      <Navigation>
        {
          buttonsState.showPreviousBtn && (
            <Button onClick={() => previous()} text="Previous" />
          )
        }
        {
          buttonsState.showNextBtn && (
            <Button onClick={() => next()} text="Next" />
          )
        }
        {
          buttonsState.showDoneBtn && (
            <Button
              onClick={() => handleDone && handleDone()}
              text={doneText}
            />
          )
        }
      </Navigation>
    )
  }

  return (
    <Wrapper
      className={classNames('c-multiStep', className)}
      onKeyDown={handleKeyDown}
    >
      {renderTracks()}
      {showNavigationTop && renderNavigation()}
      {steps[compState].component}
      {!showNavigationTop && renderNavigation()}
    </Wrapper>
  )
};

MultiStep.defaultProps = {
  doneText: 'Done',
  onDone: () => null,
  showNavigation: true,
  showNavigationTop: false,
  verticalTrack: false,
};

const getNavStyles = (index: number, length: number) => {
  let styles = [];

  for (let i = 0; i < length; i++) {
    if (i < index) {
      styles.push('-done')
    } else if (i === index) {
      styles.push('-doing')
    } else {
      styles.push('-todo')
    }
  }

  return styles;
}

const getButtonsState = (index: number, length: number) => {
  if (index > 0 && index < length - 1) {
    return {
      showDoneBtn: false,
      showPreviousBtn: true,
      showNextBtn: true,
    };
  } else if (index === 0) {
    return {
      showDoneBtn: false,
      showPreviousBtn: false,
      showNextBtn: true,
    };
  } else {
    return {
      showDoneBtn: true,
      showPreviousBtn: true,
      showNextBtn: false,
    };
  }
}

export default MultiStep;
