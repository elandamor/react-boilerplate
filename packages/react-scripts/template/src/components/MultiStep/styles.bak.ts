import styled, { css } from 'styled-components';
import theme from '../../theme';

const Wrapper = styled.div``;

export const Navigation = styled.div`
  display: flex;
  justify-content: center;
`;

export const Tracks = styled.ol`
  display: flex;
  justify-content: center;
  margin: 0;
  padding: 40px 0;

  ${(props: { verticalTrack: boolean }) => props.verticalTrack && css`
    align-items: flex-start;
    flex-direction: column;
  `};
`;

export const Track = styled.li`
  align-items: ${
    (props: { verticalTrack: boolean }) => props.verticalTrack
    ? 'flex-start' : 'center'
  };
  display: flex;
  flex-direction: column;
  justify-content: center;
  min-width: ${theme.colors.minWidths[0]}px;
  position: relative;

  .side-nav {
    &__wrap {
      align-items: center;
      cursor: pointer;
      display: flex;
      flex-direction: ${
        (props: { verticalTrack: boolean }) => props.verticalTrack
        ? 'row' : 'column'
      };
      justify-content: center;
      min-height: ${theme.colors.space[4]}px;
    }

    &__circle {
      align-items: center;
      background-color: ${({ theme }) => theme.colors.cardBorderColor};
      border-radius: 50%;
      display: flex;
      height: ${theme.colors.space[4]}px;
      justify-content: center;
      margin: ${theme.colors.space[3]}px;
      width: ${theme.colors.space[4]}px;

      &-inner {
        background-color: ${({ theme }) => theme.colors.surface};
        border-radius: 50%;
        display: block;
        height: ${theme.colors.space[3]}px;
        width: ${theme.colors.space[3]}px;
      }
    }

    &__line {
      background-color: ${({ theme }) => theme.colors.cardBorderColor};
      border-radius: 2px;
      height: ${theme.colors.space[1] / 2}px;
      position: absolute;
      right: -25%;
      visibility: hidden;
      width: 50%;

      ${(props: { verticalTrack: boolean }) => props.verticalTrack && css`
        bottom: -12.5%;
        height: 25%;
        width: ${theme.colors.space[1] / 2}px;
        right: auto;
      `};
    }
  }

  &.-done {
    .side-nav__circle {
      background-color: ${({ theme }) => theme.colors.cardBorderColor};
    }
  }

  &.-doing {
    .side-nav__circle {
      background-color: ${({ theme }) => theme.colors.primary};
    }

    .side-nav__page {
      font-weight: 800;
    }
  }

  &:not(:last-child).-done {
    .side-nav__line {
      visibility: visible;
    }
  }

  &:hover {
    .side-nav__circle {
      background-color: ${theme.colors.blacks[5]};
    }

    .-done {
      .side-nav__circle {
        background-color: ${theme.colors.blacks[5]};
      }
    }
  }
`;

export default Wrapper;
