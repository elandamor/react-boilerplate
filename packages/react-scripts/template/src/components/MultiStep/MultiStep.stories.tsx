import React from 'react';
import { storiesOf } from '@storybook/react';
import { Field, Formik, FieldArray } from 'formik';

import MultiStep from './index';
import Form from '../Form';
import Button from '../Button';
import Flex from '../Flex';

const initialValues = {
  step1: {
    name: '',
  },
  step2: {
    bio: '',
    contacts: [],
  }
};

const Step1 = (props: any) => {
  return (
    <React.Fragment>
      <Field type="text" label="Name" name="step1.name" placeholder="Name" />
    </React.Fragment>
  )
}

const Step2 = (props: any) => {
  const contacts = props.contacts ? props.contacts : undefined;

  return (
    <React.Fragment>
      <Field component="textarea" name="step2.bio" placeholder="Enter your biography" />
      <FieldArray name="step2.contacts">
        {({ push, remove }) => (
          <React.Fragment>
            {
              contacts && contacts.length > 0
              && contacts.map((contact: any, index: number) => (
                <Flex key={`contact-${index}`}>
                  <Field
                    name={`step2.contacts[${index}].name`}
                    type="text"
                    placeholder="John Doe"
                  />
                  <Field
                    name={`step2.contacts[${index}].number`}
                    type="tel"
                    placeholder="+263123456789"
                  />
                  <Button
                    onClick={() => remove(index)}
                    text="Remove"
                    iconOnly={true}
                  />
                </Flex>
              ))
            }
            <Button
              onClick={() => push({ name: '', number: '' })}
              text="Add Contact"
            />
          </React.Fragment>
        )}
      </FieldArray>
    </React.Fragment>
  )
}

storiesOf('MultiStep', module)
.add('default', () => (
  <MultiStep
    steps={[
      { name: 'Step 1', component: <div>1</div> },
      { name: 'Step 2', component: <div>2</div> },
      { name: 'Step 3', component: <div>3</div> },
    ]}
  />
))
.add('vertical', () => (
  <MultiStep
    verticalTrack
    steps={[
      { name: 'Step 1', component: <div>1</div> },
      { name: 'Step 2', component: <div>2</div> },
      { name: 'Step 3', component: <div>3</div> },
    ]}
  />
))
.add('Mulit-step form', () => (
  <Formik
    initialValues={initialValues}
    onSubmit={(values, { setSubmitting }) => {
      console.log({ values });

      setTimeout(() => setSubmitting(false), 3000);
    }}
  >
    {({ values, handleSubmit }) => (
      <Form>
        <MultiStep
          steps={[
            { name: 'Step 1', component: <Step1 {...values.step1} /> },
            { name: 'Step 2', component: <Step2 {...values.step2} /> },
          ]}
          onDone={handleSubmit}
        />
      </Form>
    )}
  </Formik>
))
