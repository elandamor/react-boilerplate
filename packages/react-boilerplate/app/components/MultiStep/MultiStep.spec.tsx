// MultiStep.spec.tsx
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import 'jest-styled-components';

import MultiStep from './index';

const Step1: React.SFC<{}> = () => <div>Step 1</div>;
const Step2: React.SFC<{}> = () => <div>Step 2</div>;
const Step3: React.SFC<{}> = () => <div>Step 3</div>;

describe('<MultiStep />', () => {
  it('should display a MultiStep', () => {
    const component = renderer
      .create(
        <MultiStep
          onSubmit={() => null}
          steps={[
            { component: <Step1 />, name: 'Step 1' },
            { component: <Step2 />, name: 'Step 2' },
            { component: <Step3 />, name: 'Step 3' },
          ]}
        />,
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });
});
