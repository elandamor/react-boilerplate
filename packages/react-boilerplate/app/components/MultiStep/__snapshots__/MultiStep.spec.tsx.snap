// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiStep /> should display a MultiStep 1`] = `
.c0 {
  padding: 24px 0;
  position: relative;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0;
  padding: 0;
  width: 100%;
}

.c1 li {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  list-style-type: none;
  position: relative;
}

.c1 li .c-icon {
  border: 4px solid red;
  border-radius: 100%;
  display: block;
  height: 48px;
  margin-bottom: 12px;
  width: 48px;
}

.c1 li.-doing .c-icon,
.c1 li.-done .c-icon {
  background-color: red;
}

.c1 li.-doing:not(:first-child):before,
.c1 li.-done:not(:first-child):before {
  background-color: red;
  content: '';
  height: 4px;
  left: -50%;
  margin-top: -2px;
  position: absolute;
  top: 24px;
  width: 100%;
}

<div
  className="c0"
  onKeyDown={[Function]}
>
  <ol
    className="c1"
  >
    <li
      className="c-step  -doing"
      onClick={[Function]}
      value={0}
    >
      <i
        className="c-icon"
      />
      <span>
        Step 1
      </span>
    </li>
    <li
      className="c-step  -todo"
      onClick={[Function]}
      value={1}
    >
      <i
        className="c-icon"
      />
      <span>
        Step 2
      </span>
    </li>
    <li
      className="c-step  -todo"
      onClick={[Function]}
      value={2}
    >
      <i
        className="c-icon"
      />
      <span>
        Step 3
      </span>
    </li>
  </ol>
  <div>
    Step 1
  </div>
  <div
    className=""
  >
    <button
      className="c-btn "
      onClick={[Function]}
    >
      Previous
    </button>
    <button
      className="c-btn "
      onClick={[Function]}
    >
      Next
    </button>
  </div>
</div>
`;
